doc/performance-testing/cloud-run/0002_stress-test-phase-2-december-2021.md

```diff
@@ -0,0 +1,122 @@
+# Stress Test Phase 2 December 2021
+
+Following the December 2021 phase 1 concurrency test we now need to scale test to ensure Runner is scaling linearly under Cloud Run.
+
+The test intends to confirm that the observations in phase 1 are linear at scale (at 50 instances).
+
+## Runner settings
+
+| Setting | Value |
+| --- | ---| 
+| Concurrency | 115 |
+| Max instances   | 50 |
+| Min instances | 50 |
+| CPU | 4 |
+| Memory | 4G |
+| Version | 3.89.0 |
+
+
+## Benchmark settings
+
+| Setting | Value |
+| --- | ---| 
+| Clients per instance | 100 |
+| Clients hatch rate   | 50 |
+| Wait time minimum | 1 |
+| Wait time maximum | 2 |
+| Requests JSON | test_benchmark_business.json |
+| Runtime | 20m |
+
+## Initial Benchmark
+
+Prior to scaling runner to 50 instances, we ran an initial benchmark against runner 1 instance to check that the results were comparable with the phase 1 test.
+
+For our phase 2 benchmark, the 99th percentile response time was slightly above our expected range of 140 to 180ms but the results were broadly similar to our phase 1 test. CPU usage peaked at 70% but for the most part stayed between 66-68%.
+
+| Stress Test |Runner Concurrency |Runner Instances | Run Time| Clients per instance | Requests per second | 99th percentile Max CPU Usage (%) | 99th percentile response time (ms) | Error rate (%) | Output |
+| --- | --- | --- | --- |--- | --- | --- | --- | --- | --- |
+| Phase 2 | 115 | 1 | 20 mins | 100  | 116.40 | 70  | 190 | 0.000 | [output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2021-12-23T11:34:57) |
+| Phase 1 | 90 | 1 | 10 mins | 90  | 104.21 | 66  | 190 | 0.001 | [output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2021-12-09T13:39:52)|
+
+
+## Results
+
```

Comment:
The Grafana visualisations are an easier way to understand the impact of the 99th% spikes at the lower percentiles than adding the raw metrics in the table. As discussed with @berroar if the 95%s and 50%s consistently spike along with the 99th then it warrants further investigation. We regularly saw spikes like this in the the 99th during the Census but it was due to outliers and anomalies rather than underlying issues.

Looking at the visualisations from these runs (particularly the re-runs of 40, 45 and 50) the 50% and 95% are not trending to the same degree as the 99th% spikes. Also, the spikes themselves (again particularly during the rerun appear to be a consequence of the ramp up and ramp down of the test load. We also saw this during Census testing.